Mathlib/Data/Matroid/Rank/Finite.lean

```diff
@@ -0,0 +1,190 @@
+
+lemma Indep.subset_finite_basis_of_subset_of_rkFin (hI : M.Indep I) (hIX : I ⊆ X)
+    (hX : M.RkFin X) (hXE : X ⊆ M.E := by aesop_mat) : ∃ J, M.Basis J X ∧ I ⊆ J ∧ J.Finite :=
+  (hI.subset_basis_of_subset hIX).imp fun _ hJ => ⟨hJ.1, hJ.2, hJ.1.finite_of_rkFin hX⟩
+
+lemma rkFin_singleton (M : Matroid α) (e : α) : M.RkFin {e} :=
+  rkFin_of_finite M (finite_singleton e)
+
+@[simp]
+lemma RkFin.empty (M : Matroid α) : M.RkFin ∅ :=
+  rkFin_of_finite M finite_empty
+
+lemma not_rkFin_superset (h : ¬M.RkFin X) (hXY : X ⊆ Y) : ¬M.RkFin Y :=
+  fun h' ↦ h (h'.subset hXY)
+
+lemma RkFin.finite_of_indep_subset (hX : M.RkFin X) (hI : M.Indep I) (hIX : I ⊆ X) : I.Finite :=
+  hI.finite_of_rkFin (hX.subset hIX)
+
+@[simp]
+lemma rkFin_ground_iff_rankFinite : M.RkFin M.E ↔ M.RankFinite := by
+  rw [RkFin, restrict_ground_eq_self]
+
+lemma rkFin_ground (M : Matroid α) [RankFinite M] : M.RkFin M.E := by
+  rwa [rkFin_ground_iff_rankFinite]
+
+lemma Indep.finite_of_subset_rkFin (hI : M.Indep I) (hIX : I ⊆ X) (hX : M.RkFin X) : I.Finite :=
+  hX.finite_of_indep_subset hI hIX
+
+lemma RkFin.closure (h : M.RkFin X) : M.RkFin (M.closure X) :=
+  let ⟨_, hI⟩ := M.exists_basis' X
+  hI.basis_closure_right.rkFin_of_finite <| hI.finite_of_rkFin h
+
+@[simp]
+lemma rkFin_closure_iff : M.RkFin (M.closure X) ↔ M.RkFin X := by
+  rw [← rkFin_inter_ground_iff (X := X)]
+  exact ⟨fun h ↦ h.subset <| M.inter_ground_subset_closure X, fun h ↦ by simpa using h.closure⟩
+
+lemma RkFin.union (hX : M.RkFin X) (hY : M.RkFin Y) : M.RkFin (X ∪ Y) := by
+  obtain ⟨I, hI, hIfin⟩ := hX.exists_finite_basis'
+  obtain ⟨J, hJ, hJfin⟩ := hY.exists_finite_basis'
+  rw [← rkFin_inter_ground_iff]
+  refine (M.rkFin_of_finite (hIfin.union hJfin)).closure.subset ?_
+  rw [closure_union_congr_left hI.closure_eq_closure,
+    closure_union_congr_right hJ.closure_eq_closure]
+  exact inter_ground_subset_closure M (X ∪ Y)
+
+lemma RkFin.rkFin_union_iff (hX : M.RkFin X) : M.RkFin (X ∪ Y) ↔ M.RkFin Y :=
+  ⟨fun h ↦ h.subset subset_union_right, fun h ↦ hX.union h⟩
+
+lemma RkFin.rkFin_diff_iff (hX : M.RkFin X) : M.RkFin (Y \ X) ↔ M.RkFin Y := by
+  rw [← hX.rkFin_union_iff, union_diff_self, hX.rkFin_union_iff]
+
+lemma RkFin.inter_right (hX : M.RkFin X) : M.RkFin (X ∩ Y) :=
+  hX.subset inter_subset_left
+
+lemma RkFin.inter_left (hX : M.RkFin X) : M.RkFin (Y ∩ X) :=
+  hX.subset inter_subset_right
+
+lemma RkFin.diff (hX : M.RkFin X) : M.RkFin (X \ Y) :=
+  hX.subset diff_subset
+
+lemma RkFin.insert (hX : M.RkFin X) (e : α) : M.RkFin (insert e X) := by
+  rw [← union_singleton]; exact hX.union (M.rkFin_singleton e)
+
+@[simp]
+lemma rkFin_insert_iff {e : α} : M.RkFin (insert e X) ↔ M.RkFin X := by
+  rw [← singleton_union, (M.rkFin_singleton e).rkFin_union_iff]
+
+@[simp]
+lemma RkFin.diff_singleton_iff : M.RkFin (X \ {e}) ↔ M.RkFin X := by
+  rw [(M.rkFin_singleton e).rkFin_diff_iff]
+
+lemma to_rkFin (M : Matroid α) [RankFinite M] (X : Set α) : M.RkFin X :=
```

Comment:
I'm not sure this name is the right one, but I don't have a better idea off the top of my head.